c7n/resources/cloudfront.py

```diff
@@ -181,9 +182,130 @@ def process(self, resources, event=None):
+                    'Non-accessible bucket: %s' % target_bucket})
+            if e.response['Error']['Code'] == 'NoSuchBucket':
+                self.log.debug({'state': 'error', 'reason':
+                    'Non-existent bucket: %s' % target_bucket})
+
+        return None
+
+    def process_custom_target(self, r, target_origin, accounts):
+        try:
+            if accounts and target_origin not in accounts:
+                self.log.debug("Custom origin %s not in origins whitelist." % target_origin)
+            else:
+                self.log.debug("Custom origin %s found in origins whitelist." % target_origin)
+                r['c7n:target-origin'] = True
+                return r
+
+        except ClientError as e:
+            self.log.debug(e)
+
+        return None
+
+    def process_resource_origins(self, resources, accounts):
+        results = []
+
+        if 'S3' in self.data['origins'] or 'S3-CUSTOM' in self.data['origins']:
+            client = local_session(self.manager.session_factory).client(
+                's3', region_name=self.manager.config.region)
+
+        for r in resources:
+            for x in r['Origins']['Items']:
+                if 'S3' in self.data['origins'] and 'S3OriginConfig' in x:
+                    target_bucket = x['DomainName'].split('.', 1)[0]
+                    resource = self.process_s3_target(r, target_bucket, client, accounts)
+                    results += [resource] if resource else []
+                elif 'S3-CUSTOM' in self.data['origins'] and 'CustomOriginConfig' in x \
+                        and self.is_s3_domain(x['DomainName']):
+                    target_bucket = self.get_s3_domain(x['DomainName'])
+                    resource = self.process_s3_target(r, target_bucket, client, accounts)
+                    results += [resource] if resource else []
+                elif 'CUSTOM' in self.data['origins'] and 'CustomOriginConfig' in x:
+                    target_origin = x['DomainName']
+                    resource = self.process_custom_target(r, target_origin, accounts)
+                    results += [resource] if resource else []
+
+        return results
+
+    def is_s3_domain(self, domain_name):
+        if domain_name.endswith('.s3.amazonaws.com'):
+            return True
+        else:
+            return False
+
+    def get_s3_domain(self, domain_name):
+        if domain_name.endswith('.s3.amazonaws.com'):
+            return domain_name[:-len('.s3.amazonaws.com')]
+
+    def process(self, resources, event=None):
+
+        accounts = set(self.data.get('accounts', ()))
+        if 'accounts_from' in self.data:
+            values = ValuesFrom(self.data['accounts_from'], self.manager).get_values()
+            accounts = accounts.union(values)
+
+        if accounts:
+            results = self.process_resource_origins(resources, accounts)
+        else:
+            raise ValueError('Accounts whitelist is empty. Policy failure.')
```

Comment:
this should be done in the validate method of this filter or in the schema itself